### DISCUSSION TITLE
how to use useFetch to get data from backend in composable/useXXX.ts?

### DISCUSSION THREAD

**Discussion by JiatLn** [nuxt/nuxt#17058] [Questions] [answered]
- useArticle.ts
```ts
export function useArticleList(categoryId: Ref<number>) {
  const articleList = ref<ArticleItem[]>([])
  const loading = ref<boolean>(false)

  async function fetchData() {
    loading.value = true
    const { data } = await useFetch<FetchRow<ArticleItem>>('/api/articleList', {
      params: {
        pageNum: 1,
        pageSize: 6,
        cate: categoryId.value,
      },
    })
    if (data.value.code === OK_CODE) {
      articleList.value = data.value.rows.slice(0, 6)
    }
    loading.value = false
  }

  fetchData()
  watch(categoryId, fetchData)
  return {
    loading,
    articleList,
    fetchData,
  }
}
```

- index.vue

```ts
const ida = ref<number>(172)
const idb = ref<number>(173)
const { articleList: aaa } = useArticleList(ida)
const { articleList: bbb } = useArticleList(idb)
```

- server/api/articleList.ts

```ts
export default defineEventHandler(async (event) => {
  const config = useRuntimeConfig().public
  const q = useQuery(event.req)
  const url = `${config.apiBase}/article/front/AllArticleList/${q.cate}`
  const data = await $fetch(url, {
    method: 'GET',
    params: {
      ...q,
    },
  })
  return data
})
```

when i use the code above, i got the same data from backend that the `cate` is 172, the 173 is not work :(
the result is that `bbb` is same as `aaa`

**Comment by ShaggyTech** [comment, marked as answer]
Edit: First I would console.log(categoryId.value) in the watch callback to make sure the value is what you expect. 

You can use `key` in useFetch options.  

> key: a unique key to ensure that data fetching can be properly de-duplicated across requests, if not provided, it will be generated based on the url and fetch options.

See [https://v3.nuxtjs.org/api/composables/use-fetch#params](https://v3.nuxtjs.org/api/composables/use-fetch#params)

Although I would think that the fetch options params changing would count as 'fetch options' changing according to: 
> if not provided, it will be generated based on the url and fetch options.

Example using categoryId.value as useFetch key:

```ts
const { data } = await useFetch<FetchRow<ArticleItem>>('/api/articleList', {
  params: {
    pageNum: 1,
    pageSize: 6,
    cate: categoryId.value,
  },
  key: categoryId.value
})
```

**Reply by JiatLn** [reply]
thx, i add a key and it work!
